### DISCUSSION TITLE
How to change instrument in a midi?

### DISCUSSION THREAD

**Discussion by clumsyroger** [mido/mido#436] [Q&A]
One MIDI file has only one instrument, the piano.I want to change the  instrument of this midi to other instruments, such as the violin, and then remove all the notes that are not in the range of the violin. What should I do with this mido? Could anyone give me some advice? Thanks a lot!

**Comment by rdoursenaud** [comment]
For changing instruments, look into MIDI program changes.
To filter notes by range, you’ll have to code that yourself ;)

**Comment by page200** [comment]
I guess what's meant are things like `Message('program_change', channel=0, program=0, time=0)`. The meaning of the arguments seem not well documented in `mido`, and the relationship to instruments is not mentioned in the documentation.

Moreover, the `channel` variable in `'note_on'` events should be chosen accordingly.

**Comment by rdoursenaud** [comment]
```MIDI``` doesn’t specify any Program Number/Instrument relationship directly. This depends on the target instrument, hence why there is currently no helper in ```mido```.

There are several optional standard and proprietary specifications sufficiently popular (General MIDI 1 aka GM1, General MIDI 2 aka GM2, General MIDI Lite, GS, XG…) that we may want to support at a later time.

My guess is that the sequencer you’re using is displaying ```General MIDI 1``` instrument names. You can find the mapping in the [specification](https://www.midi.org/specifications/midi1-specifications/general-midi-specifications/general-midi-1).

Also, there is no such thing as an instrument per channel in the ```Standard MIDI File``` (aka SMF) specification. Sequencers usually derives them from the first ```Program Change``` message of the Track (Itself often but not always mapped to a ```MIDI Channel```).

**Reply by page200** [reply]
> Also, there is no such thing as an instrument per channel in the `Standard MIDI File` (aka SMF) specification. Sequencers usually derives them from the first `Program Change` message of the Track (Often but not always mapped to a `MIDI Channel`).

What is often but not always mapped to a `MIDI Channel`: the `Program Change` message (then what value is stored in that MIDI message where usually the channel number is stored?) or the track (then where in the `mid` object from `mido` will that message be stored, if usually that object consists of MIDI tracks?)? Do you have an example MIDI file where it's not mapped to a `MIDI Channel`?

**Reply by rdoursenaud** [reply]
> What is often but not always mapped to a `MIDI Channel`: the `Program Change` message [...] or the track [...]

The `MIDIFile` `Track`.

Depending on the MIDI File `format` (sometimes inaccurately referred to as `type` including in `mido`), there could only be:
- a single track (Format 0)
- any number of synchronous tracks (Format 1)
- any number of independent/asynchronous tracks (Format 2)

All tracks in all formats can contain `Channel Messages` like the `Program Change` which require a channel to address.

The gist of what I was trying to convey is:
There isn't a one-to-one relationship between MIDI Channels and MIDI File Tracks.
Some files are structured that way but it isn't a requirement. Some DAW Sequencers gets that wrong which results in users making false assumptions about the format and sometimes causes erroneous playback.

**Reply by page200** [reply]
> The `MIDIFile` `Track` [is often mapped to a `MIDI Channel`].

The output of `mido.MidiFile()` makes me disagree. There, each track contains events, and most events are mapped to some channel. A large percentage of MIDI files contain at least one track that contains messages for several different channels.

And I don't see any other relationship between tracks and channels in the output of `mido.MidiFile()`.

**Reply by rdoursenaud** [reply]
There isn't. I was talking about MIDI files floating around. More often then not, their author group channel messages on a track per channel. This is the way most sequencers operate in their default configuration. It is not a property of the file format but a remark on how it is often used. I thought I needed to clarify since users often makes incorrect assumptions about the format inferring from the way it's used in the wild. Hope that clears things up.

**Reply by rdoursenaud** [reply]
Confusion often raises from the fact that some sequencers insists on mapping *their* tracks to a MIDI channel and/or instrument which is not how the file format tracks necessarily operates.
